core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java

```diff
@@ -0,0 +1,378 @@
+
+        source.subscribe(innerSubscriber, executor, options);
+    }
+
+    @Override
+    public void abort() {
+        source.abort();
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        source.abort(cause);
+    }
+
+    private static final class FlatMapAggregatingSubscriber<T, U> implements Subscriber<T>, Subscription {
+        private final int maxConcurrency;
+
+        private final Subscriber<? super U> downstream;
+        private final Function<T, StreamMessage<U>> function;
+        private final EventExecutor executor;
+        private final Set<FlatMapSubscriber<T, U>> childSubscribers;
+        private final Queue<U> buffer;
+        private final CompletableFuture<Void> completionFuture;
+        private final SubscriptionOption[] options;
+
+        @Nullable
+        private volatile Subscription upstream;
+        private volatile boolean canceled;
+
+        private long requestedByDownstream;
+        private int pendingSubscriptions;
+        private boolean completing;
+
+        FlatMapAggregatingSubscriber(Subscriber<? super U> downstream,
+                                     Function<T, StreamMessage<U>> function,
+                                     EventExecutor executor,
+                                     int maxConcurrency,
+                                     CompletableFuture<Void> completionFuture,
+                                     SubscriptionOption... options) {
+            requireNonNull(downstream, "downstream");
+            requireNonNull(function, "function");
+            requireNonNull(executor, "executor");
+            requireNonNull(completionFuture, "completionFuture");
+
+            this.downstream = downstream;
+            this.function = function;
+            this.executor = executor;
+            this.maxConcurrency = maxConcurrency;
+            this.completionFuture = completionFuture;
+            this.options = options;
+
+            childSubscribers = new HashSet<>();
+            buffer = new ArrayDeque<>();
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+
+            upstream = subscription;
+            downstream.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+
+            if (canceled) {
+                StreamMessageUtil.closeOrAbort(item);
+                return;
+            }
+
+            pendingSubscriptions++;
+            final StreamMessage<U> newStreamMessage = function.apply(item);
+            newStreamMessage.subscribe(new FlatMapSubscriber<>(this), executor, options);
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            requireNonNull(cause, "cause");
+
+            if (canceled) {
+                return;
+            }
+            canceled = true;
+
```

Comment:
I guess whatever is in the `buffer` needs to be cleaned up/closed using `StreamMessageUtil.closeOrAbort` and also `whenComplete` needs to be completed exceptionally.

Same goes for when `FlatMapAggregatingSubscriber#cancel` is called.